### ISSUE TITLE
Poor directory snapshotting performance for source directory sets with many entries

### ISSUE THREAD

**Issue by DanielThomas** [gradle/gradle#26757]
### Current Behavior

We have a project with 342 source directories, interface files for messaging, and we're seeing very poor `processResources` task performance, both input snapshotting and at task runtime.

<img width="874" alt="Screenshot 2023-10-18 at 1 42 37 pm" src="https://github.com/gradle/gradle/assets/1479220/0e1aa8b4-0311-45ea-b39d-3dfa455db83a">

Appears the problem is `getSourceTrees`. In the case of snapshotting, it's called for every file tree visit:

```
getSourceTrees:245, DefaultSourceDirectorySet (org.gradle.api.internal.file)
getSourceTrees:-1, DefaultSourceDirectorySet_Decorated (org.gradle.api.internal.file)
visitChildren:275, DefaultSourceDirectorySet (org.gradle.api.internal.file)
getSourceCollections:127, CompositeFileCollection (org.gradle.api.internal.file)
getSourceCollections:54, CompositeFileTree (org.gradle.api.internal.file)
getSourceCollections:-1, DefaultSourceDirectorySet_Decorated (org.gradle.api.internal.file)
contains:56, CompositeFileCollection (org.gradle.api.internal.file)
lambda$new$e8f07a1$1:71, DefaultSourceSet (org.gradle.api.internal.tasks)
isSatisfiedBy:-1, DefaultSourceSet$$Lambda/0x00000070019db6a0 (org.gradle.api.internal.tasks)
isSatisfiedBy:48, OrSpec (org.gradle.api.specs)
isSatisfiedBy:36, NotSpec (org.gradle.api.specs)
findUnsatisfiedSpec:67, AndSpec (org.gradle.api.specs)
isSatisfiedBy:51, AndSpec (org.gradle.api.specs)
lambda$getAsDirectoryWalkerPredicate$1:65, PatternSetSnapshottingFilter (org.gradle.internal.fingerprint.impl)
test:-1, PatternSetSnapshottingFilter$$Lambda/0x0000007001b02d28 (org.gradle.internal.fingerprint.impl)
shouldVisit:475, DirectorySnapshotter$PathVisitor (org.gradle.internal.snapshot.impl)
shouldVisitDirectory:398, DirectorySnapshotter$PathVisitor (org.gradle.internal.snapshot.impl)
doPreVisitDirectory:295, DirectorySnapshotter$PathVisitor (org.gradle.internal.snapshot.impl)
preVisitDirectory:117, DirectorySnapshotterStatistics$CollectingFileVisitor (org.gradle.internal.snapshot.impl)
preVisitDirectory:106, DirectorySnapshotterStatistics$CollectingFileVisitor (org.gradle.internal.snapshot.impl)
walkFileTree:2792, Files (java.nio.file)
snapshot:119, DirectorySnapshotter (org.gradle.internal.snapshot.impl)
lambda$snapshot$8:155, DefaultFileSystemAccess (org.gradle.internal.vfs.impl)
snapshot:-1, DefaultFileSystemAccess$$Lambda/0x000000700147c028 (org.gradle.internal.vfs.impl)
store:84, AbstractVirtualFileSystem (org.gradle.internal.vfs.impl)
snapshot:145, DefaultFileSystemAccess (org.gradle.internal.vfs.impl)
lambda$read$7:129, DefaultFileSystemAccess (org.gradle.internal.vfs.impl)
get:-1, DefaultFileSystemAccess$$Lambda/0x0000007001af8dc0 (org.gradle.internal.vfs.impl)
orElseGet:364, Optional (java.util)
lambda$readSnapshotFromLocation$9:187, DefaultFileSystemAccess (org.gradle.internal.vfs.impl)
get:-1, DefaultFileSystemAccess$$Lambda/0x000000700147b5e0 (org.gradle.internal.vfs.impl)
guardByKey:221, DefaultFileSystemAccess$StripedProducerGuard (org.gradle.internal.vfs.impl)
lambda$readSnapshotFromLocation$10:184, DefaultFileSystemAccess (org.gradle.internal.vfs.impl)
get:-1, DefaultFileSystemAccess$$Lambda/0x000000700147b3c0 (org.gradle.internal.vfs.impl)
orElseGet:364, Optional (java.util)
readSnapshotFromLocation:184, DefaultFileSystemAccess (org.gradle.internal.vfs.impl)
read:126, DefaultFileSystemAccess (org.gradle.internal.vfs.impl)
visitFileTree:76, DefaultFileCollectionSnapshotter$SnapshottingVisitor (org.gradle.internal.fingerprint.impl)
visitFileTree:125, FileTreeAdapter$1 (org.gradle.api.internal.file.collections)
visitStructure:106, DirectoryFileTree (org.gradle.api.internal.file.collections)
visitContents:121, FileTreeAdapter (org.gradle.api.internal.file.collections)
visitStructure:366, AbstractFileCollection (org.gradle.api.internal.file)
lambda$visitContents$0:133, CompositeFileCollection (org.gradle.api.internal.file)
accept:-1, CompositeFileCollection$$Lambda/0x0000007001a9b7d8 (org.gradle.api.internal.file)
lambda$visitChildren$0:67, FilteredFileTree (org.gradle.api.internal.file)
accept:-1, FilteredFileTree$$Lambda/0x0000007001af8000 (org.gradle.api.internal.file)
visitFileTree:84, FileCollectionBackedFileTree$1 (org.gradle.api.internal.file)
visitFileTree:111, FileCollectionBackedFileTree$2 (org.gradle.api.internal.file)
visitFileTree:125, FileTreeAdapter$1 (org.gradle.api.internal.file.collections)
visitStructure:106, DirectoryFileTree (org.gradle.api.internal.file.collections)
visitContents:121, FileTreeAdapter (org.gradle.api.internal.file.collections)
visitStructure:366, AbstractFileCollection (org.gradle.api.internal.file)
lambda$visitContents$0:133, CompositeFileCollection (org.gradle.api.internal.file)
accept:-1, CompositeFileCollection$$Lambda/0x0000007001a9b7d8 (org.gradle.api.internal.file)
visitChildren:276, DefaultSourceDirectorySet (org.gradle.api.internal.file)
visitContents:133, CompositeFileCollection (org.gradle.api.internal.file)
visitStructure:366, AbstractFileCollection (org.gradle.api.internal.file)
lambda$visitContents$0:133, CompositeFileCollection (org.gradle.api.internal.file)
accept:-1, CompositeFileCollection$$Lambda/0x0000007001a9b7d8 (org.gradle.api.internal.file)
add:67, UnpackingVisitor (org.gradle.api.internal.file.collections)
visitContents:419, DefaultConfigurableFileCollection$UnresolvedItemsCollector (org.gradle.api.internal.file.collections)
visitChildren:328, DefaultConfigurableFileCollection (org.gradle.api.internal.file.collections)
visitContents:133, CompositeFileCollection (org.gradle.api.internal.file)
visitStructure:366, AbstractFileCollection (org.gradle.api.internal.file)
visitContents:96, FileCollectionBackedFileTree (org.gradle.api.internal.file)
visitContentsAsFileTrees:76, FileCollectionBackedFileTree (org.gradle.api.internal.file)
visitChildren:67, FilteredFileTree (org.gradle.api.internal.file)
visitContents:133, CompositeFileCollection (org.gradle.api.internal.file)
visitStructure:366, AbstractFileCollection (org.gradle.api.internal.file)
lambda$visitContents$0:133, CompositeFileCollection (org.gradle.api.internal.file)
accept:-1, CompositeFileCollection$$Lambda/0x0000007001a9b7d8 (org.gradle.api.internal.file)
add:67, UnpackingVisitor (org.gradle.api.internal.file.collections)
add:92, UnpackingVisitor (org.gradle.api.internal.file.collections)
visitChildren:285, DefaultFileCollectionFactory$ResolvingFileCollection (org.gradle.api.internal.file)
visitContents:133, CompositeFileCollection (org.gradle.api.internal.file)
visitStructure:366, AbstractFileCollection (org.gradle.api.internal.file)
lambda$visitContents$0:133, CompositeFileCollection (org.gradle.api.internal.file)
accept:-1, CompositeFileCollection$$Lambda/0x0000007001a9b7d8 (org.gradle.api.internal.file)
visitChildren:48, PropertyFileCollection (org.gradle.api.internal.tasks)
visitContents:133, CompositeFileCollection (org.gradle.api.internal.file)
visitStructure:366, AbstractFileCollection (org.gradle.api.internal.file)
snapshot:47, DefaultFileCollectionSnapshotter (org.gradle.internal.fingerprint.impl)
visitInputFileProperty:133, DefaultInputFingerprinter$InputCollectingVisitor (org.gradle.internal.execution.impl)
visitInputFileProperty:131, SkipEmptyWorkStep$1 (org.gradle.internal.execution.steps)
visitRegularInputs:328, TaskExecution (org.gradle.api.internal.tasks.execution)
lambda$fingerprintPrimaryInputs$2:127, SkipEmptyWorkStep (org.gradle.internal.execution.steps)
accept:-1, SkipEmptyWorkStep$$Lambda/0x0000007001af3d98 (org.gradle.internal.execution.steps)
fingerprintInputProperties:63, DefaultInputFingerprinter (org.gradle.internal.execution.impl)
fingerprintPrimaryInputs:118, SkipEmptyWorkStep (org.gradle.internal.execution.steps)
execute:77, SkipEmptyWorkStep (org.gradle.internal.execution.steps)
execute:53, SkipEmptyWorkStep (org.gradle.internal.execution.steps)
execute:32, RemoveUntrackedExecutionStateStep (org.gradle.internal.execution.steps)
execute:21, RemoveUntrackedExecutionStateStep (org.gradle.internal.execution.steps)
execute:38, MarkSnapshottingInputsStartedStep (org.gradle.internal.execution.steps.legacy)
execute:36, LoadPreviousExecutionStateStep (org.gradle.internal.execution.steps)
execute:23, LoadPreviousExecutionStateStep (org.gradle.internal.execution.steps)
execute:75, CleanupStaleOutputsStep (org.gradle.internal.execution.steps)
execute:41, CleanupStaleOutputsStep (org.gradle.internal.execution.steps)
lambda$execute$2:66, ExecuteWorkBuildOperationFiringStep (org.gradle.internal.execution.steps)
get:-1, ExecuteWorkBuildOperationFiringStep$$Lambda/0x0000007001af6250 (org.gradle.internal.execution.steps)
orElseGet:364, Optional (java.util)
execute:66, ExecuteWorkBuildOperationFiringStep (org.gradle.internal.execution.steps)
execute:38, ExecuteWorkBuildOperationFiringStep (org.gradle.internal.execution.steps)
lambda$execute$0:32, AssignWorkspaceStep (org.gradle.internal.execution.steps)
executeInWorkspace:-1, AssignWorkspaceStep$$Lambda/0x0000007001af5df0 (org.gradle.internal.execution.steps)
withWorkspace:293, TaskExecution$4 (org.gradle.api.internal.tasks.execution)
execute:30, AssignWorkspaceStep (org.gradle.internal.execution.steps)
execute:21, AssignWorkspaceStep (org.gradle.internal.execution.steps)
execute:37, IdentityCacheStep (org.gradle.internal.execution.steps)
execute:27, IdentityCacheStep (org.gradle.internal.execution.steps)
execute:47, IdentifyStep (org.gradle.internal.execution.steps)
execute:34, IdentifyStep (org.gradle.internal.execution.steps)
execute:64, DefaultExecutionEngine$1 (org.gradle.internal.execution.impl)
executeIfValid:145, ExecuteActionsTaskExecuter (org.gradle.api.internal.tasks.execution)
execute:134, ExecuteActionsTaskExecuter (org.gradle.api.internal.tasks.execution)
execute:46, FinalizePropertiesTaskExecuter (org.gradle.api.internal.tasks.execution)
execute:51, ResolveTaskExecutionModeExecuter (org.gradle.api.internal.tasks.execution)
execute:57, SkipTaskWithNoActionsExecuter (org.gradle.api.internal.tasks.execution)
execute:74, SkipOnlyIfTaskExecuter (org.gradle.api.internal.tasks.execution)
execute:36, CatchExceptionTaskExecuter (org.gradle.api.internal.tasks.execution)
executeTask:77, EventFiringTaskExecuter$1 (org.gradle.api.internal.tasks.execution)
call:55, EventFiringTaskExecuter$1 (org.gradle.api.internal.tasks.execution)
call:52, EventFiringTaskExecuter$1 (org.gradle.api.internal.tasks.execution)
execute:204, DefaultBuildOperationRunner$CallableBuildOperationWorker (org.gradle.internal.operations)
execute:199, DefaultBuildOperationRunner$CallableBuildOperationWorker (org.gradle.internal.operations)
execute:66, DefaultBuildOperationRunner$2 (org.gradle.internal.operations)
execute:59, DefaultBuildOperationRunner$2 (org.gradle.internal.operations)
execute:157, DefaultBuildOperationRunner (org.gradle.internal.operations)
execute:59, DefaultBuildOperationRunner (org.gradle.internal.operations)
call:53, DefaultBuildOperationRunner (org.gradle.internal.operations)
call:78, DefaultBuildOperationExecutor (org.gradle.internal.operations)
execute:52, EventFiringTaskExecuter (org.gradle.api.internal.tasks.execution)
execute:42, LocalTaskNodeExecutor (org.gradle.execution.plan)
execute:331, DefaultTaskExecutionGraph$InvokeNodeExecutorsAction (org.gradle.execution.taskgraph)
execute:318, DefaultTaskExecutionGraph$InvokeNodeExecutorsAction (org.gradle.execution.taskgraph)
lambda$execute$0:314, DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction (org.gradle.execution.taskgraph)
run:-1, DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction$$Lambda/0x0000007001ad9520 (org.gradle.execution.taskgraph)
with:80, CurrentBuildOperationRef (org.gradle.internal.operations)
execute:314, DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction (org.gradle.execution.taskgraph)
execute:303, DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction (org.gradle.execution.taskgraph)
execute:463, DefaultPlanExecutor$ExecutorWorker (org.gradle.execution.plan)
run:380, DefaultPlanExecutor$ExecutorWorker (org.gradle.execution.plan)
onExecute:64, ExecutorPolicy$CatchAndRecordFailures (org.gradle.internal.concurrent)
run:47, AbstractManagedExecutor$1 (org.gradle.internal.concurrent)
runWorker:1144, ThreadPoolExecutor (java.util.concurrent)
run:642, ThreadPoolExecutor$Worker (java.util.concurrent)
runWith:1596, Thread (java.lang)
run:1583, Thread (java.lang)
```

And repeatedly doing this work for these calls to `CompositeFileCollection.contains` would be excessive in any case, it's worse because a two calls to `getBooleanAttributes0` (i.e. `stat`) is required for every invocation, which is quite expensive because of all of the allocations that need to happen for Java -> Platform string conversion:

https://github.com/gradle/gradle/blob/0e65f7703845b0f47ef7cfcbfe82a3922bd689e3/subprojects/core/src/main/java/org/gradle/api/internal/file/DefaultSourceDirectorySet.java#L252-L254

<img width="1350" alt="Screenshot 2023-10-18 at 10 27 38 am" src="https://github.com/gradle/gradle/assets/1479220/083b900e-7c18-4bb0-af74-92af4f3f92a2">

Allocation pressure and GCs spike during these periods, which is why the overhead is particularly high. Individual calls to `getBooleanAttributes0` are quite quick.

### Expected Behavior

Seems like you could make an immutable copy that doesn't have this behaviour as part of walking the collection. The exists/isDirectory checks also strike me as something that could be an assertion, rather than a runtime check.

### Steps to Reproduce

Configure many `srcDirs` say ~300  for each source set:

```
        project.sourceSets.main.java.srcDirs += relativeDirectory + '/src/main/java'
        project.sourceSets.main.proto.srcDirs += relativeDirectory + '/src/main/proto'
        project.sourceSets.main.resources.srcDirs += relativeDirectory + '/src/main/resources'
```

Run `processResources` and note the performance.

### Gradle version

Gradle 8.4

**Comment by DanielThomas** [comment]
Appears the reason we hit this so heavily, is the `main` source set is used as an exclude when walking resource file snapshots:

<img width="1705" alt="Screenshot 2023-10-19 at 9 47 41 am" src="https://github.com/gradle/gradle/assets/1479220/8cfb6a50-2815-42ec-bc5f-b6a3fa6c723f">

**Comment by ov7a** [comment]
Thank you for providing a valid report.

The issue is in the backlog of the relevant team, but this area of Gradle is currently not a focus one, so it might take a while before a fix is made.

**Comment by cobexer** [comment]
Fixed by #26873
